core/storage/user/gae_models.py

```diff
@@ -97,6 +97,133 @@ def get_by_normalized_username(cls, normalized_username):
+            IncompleteExplorationUserModel. A instance of the
+            IncompleteExplorationUserModel if present or it creates a new one.
+        """
+        incomplete_exploration_user_model = IncompleteExplorationUserModel.get(
+            self.id, exploration_id)
+
+        if not incomplete_exploration_user_model:
+            incomplete_exploration_user_model = (
+                IncompleteExplorationUserModel.create(self.id, exploration_id))
+
+        return incomplete_exploration_user_model
+
+
+class IncompleteExplorationUserModel(base_models.BaseModel):
+    """Stores the "last playthrough" information of the exploration since the
+    user left it partially completed.
+
+    Instances of this class have keys of the form
+    [user_id].[exploration_id]
+    """
+    # The user id.
+    user_id = ndb.StringProperty(required=True, indexed=True)
+    # The exploration id.
+    exploration_id = ndb.StringProperty(required=True, indexed=True)
+    # The version of the exploration last played by the user.
+    version_last_played = ndb.IntegerProperty(default=None)
+    # The time since the user last played the exploration.
+    time_last_played_msec = ndb.FloatProperty(default=None)
+    # The state at which the learner left the exploration when he/she last
+    # played it.
+    last_state_played = ndb.StringProperty(default=None)
+
+    @classmethod
+    def _generate_id(cls, user_id, exploration_id):
+        return '%s.%s' % (user_id, exploration_id)
+
+    @classmethod
+    def create(cls, user_id, exploration_id):
+        """Creates a new IncompleteExplorationUserModel instance and returns it.
+
+        Args:
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+
+        Returns:
+            IncompleteExplorationUserModel. The newly created
+            IncompleteExplorationUserModel instance.
+        """
+        instance_id = cls._generate_id(user_id, exploration_id)
+        return cls(
+            id=instance_id, user_id=user_id, exploration_id=exploration_id)
+
+    @classmethod
+    def get(cls, user_id, exploration_id):
+        """Gets the IncompleteExplorationUserModel for the given user and
+        exploration id.
+
+        Args:
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+
+        Returns:
+            IncompleteExplorationUserModel. The IncompleteExplorationUserModel
+            instance which matches with the given user_id and exploration_id.
+        """
+        instance_id = cls._generate_id(user_id, exploration_id)
+        return super(IncompleteExplorationUserModel, cls).get(
+            instance_id, strict=False)
+
+    def update_last_played_information(self, time_last_played_msec,
+                                       version_last_played,
+                                       last_state_played):
+        """Updates the last playthrough information of the user.
```

Comment:
Hi @Arunabh98, you asked: 

> Yes, currently it saves the last played status of the exploration.Not sure whether this should happen though. Saving an older information can lead to potential problems such a saving a state that doesn't exist in the new version. So I think it would be best to save only if the version is higher, right?

I don't know; this is something you should probably discuss with @jaredsilver and @rachelwchen!